File: CSharp/ShopTest/Customer/CustomerTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;

namespace ShopTest.Customer {
    using Shop.Customers;
    using Shop.Items;
    using Shop;
    public class CustomerTest {
        [Fact]
        public void Funds_EqualsGivenMoney() {
            Customer customer = new Customer(100);
            Assert.Equal(100, customer.Funds);

            Customer customer2 = new Customer(150);
            Assert.Equal(150, customer2.Funds);
        }

        [Fact]
        public void Buy_FromShop_LowersFunds() {
            Shop shop = new Shop();
            Item i = new Item("Black Shirt", 15);
            shop.AddItem(i);

            Customer customer = new Customer(100);
            customer.Buy(shop, "Black Shirt");

            Assert.Equal(85, customer.Funds);
        }
        [Fact]
        public void Buy_ReceivesItem() {
            Shop shop = new Shop();
            Item i = new Item("Black Shirt", 15);
            shop.AddItem(i);

            Customer customer = new Customer(100);

            Assert.Empty(customer.Items);

            customer.Buy(shop, "Black Shirt");

            Assert.Single(customer.Items);
            Assert.Equal(i, customer.Items[0]);
        }
        [Fact]
        public void Return_ReturnsItem() {
            Shop shop = new Shop();
            Item i = new Item("Black Shirt", 15);
            shop.AddItem(i);

            Customer customer = new Customer(100);

            customer.Buy(shop, "Black Shirt");
            Assert.Single(customer.Items);
            Assert.Equal(85, customer.Funds);

            customer.Return(shop, customer.Items[0]);
    
            Assert.Empty(customer.Items);
            Assert.Equal(100, customer.Funds);
        }
    }
}

File: Java/TddShop/TddShopTest/src/Customer/CustomerTest.java
package Customer;

import org.junit.Assert;
import org.junit.Test;
import static org.junit.Assert.assertEquals;
import Shop.*;
import Item.*;

public class CustomerTest {
    @Test
    public void Funds_EqualsGiven(){
        Wallet w = new Wallet(101);
        Customer customer = new Customer(w);
        assertEquals(101, customer.Wallet().Funds);

        Wallet w2 = new Wallet(11);
        Customer customer2 = new Customer(w2);
        assertEquals(11, customer2.Wallet().Funds);
    }
    @Test
    public void Buy_FromShop_LowersFunds() throws NotEnoughFundsException, ItemNotAvailableException, CannotReturnException{
        Shop shop = new Shop();
        Item i = new Item("Black Shirt", 15);
        ItemBatch b = new ItemBatch(i, 1);
        shop.Add(b);

        Wallet w = new Wallet(100);
        Customer c = new Customer(w);
        c.Buy(shop, "Black Shirt");

        Assert.assertEquals(85, c.Wallet().Funds);
    }
    @Test
    public void Buy_ReceivesItem() throws NotEnoughFundsException, ItemNotAvailableException, CannotReturnException{
        Shop shop = new Shop();
        Item i = new Item("Black Shirt", 15);
        ItemBatch batch = new ItemBatch(i,1);
        shop.Add(batch);

        Customer customer = new Customer(new Wallet(100));
        assertEquals(0, customer.Items().size());

        customer.Buy(shop, "Black Shirt");

        assertEquals(1, customer.Items().size());;
        assertEquals(i, customer.Items().get(0));
    }
    @Test
    public void Return_ReturnsItemAndMoney() throws ItemNotAvailableException, NotEnoughFundsException, CannotReturnException{
        Shop shop = new Shop();
        Item i = new Item("Black Shirt", 15);
        ItemBatch batch = new ItemBatch(i, 1);
        shop.Add(batch);

        Customer customer = new Customer(new Wallet(100));

        customer.Buy(shop, "Black Shirt");
        Assert.assertEquals(1, customer.Items().size());
        Assert.assertEquals(85, customer.Wallet().Funds);

        customer.Return(shop, customer.Items().get(0));

        assertEquals(0, customer.Items().size());
        assertEquals(100, customer.Wallet().Funds);
    }
}
